File: home_endpoint/func.py
import fdk
import json
import sys
import requests

"""
  input: 
  output: 
"""
def handler(ctx, data=None, loop=None):
    sys.stderr.write("------- RECEIVING GOOGLE HOME REQUEST! -------")
    if data and len(data) > 0:
        body = json.loads(data)
        sys.stderr.write(json.dumps(body))

    searchPhrase = body["queryResult"]["queryText"]
    sys.stderr.write("search phrase from Google Home --> " + searchPhrase)

    process_url = "https://carimura.ngrok.io/t/giphyfn/flow"
    headers = {'Content-Type': 'text/plain'}

    # this kills my response back to Google.. timing out??
    response = requests.post(process_url, headers=headers, data=searchPhrase)
    sys.stderr.write("status_code --> " + response.status_code)
    
    resp = {
            "payload": {
                "google": {
                "expectUserResponse": "false",
                "richResponse": {
                    "items": [
                    {
                        "simpleResponse": {
                        "textToSpeech": "OK I've invoked the Flow function for you."
                        }
                    }
                    ]
                }
                }
            }
            }

    sys.stderr.write("-------// ENDING HOME_ENDPOINT -------")
    return resp

if __name__ == "__main__":
    fdk.handle(handler)

File: home_endpoint/func.js
const fdk = require('@fnproject/fdk');
const request = require('request');
const FN_SERVER_URL = process.env.FN_SERVER_URL

/* 
  This function takes a webhook from Google Home and triggers
  the flow function.

  input: [json] some long json that Google Home sends
  action: triggers the Flow function
  output: [json] response that Google Home expects
*/

fdk.handle(function (input) {
  return callFlow(input);
});

async function callFlow(input) {
  var searchPhrase = "cats"

  if (input.queryResult && input.queryResult.parameters && input.queryResult.parameters.searchPhrase) {
    searchPhrase = input.queryResult.parameters.searchPhrase
  }
  console.log("searchPhrase --> " + searchPhrase)

  var url = FN_SERVER_URL + "/t/giphyfn/flow"

  return new Promise((resolve, reject) => {
    request.post(url, {
      headers: { 'content-type': 'application/x-www-form-urlencoded' },
      body: searchPhrase
    },
      async function (err, resp, body) {
        if (err) { return console.log(err) }
        r = { 'fulfillmentText': `OK I've triggered the flow function with search term ${searchPhrase}` }
        resolve(r)
      }
    );
  });

}


